art/estimators/object_tracking/pytorch_goturn.py

```diff
@@ -0,0 +1,626 @@
+            y_pred = self.track(x=x_i, y_init=y_init[i])
+            gt_bb = labels_t[i]["boxes"]
+            loss = torch.nn.L1Loss(size_average=False)(y_pred.float(), gt_bb.float())
+            loss_list.append(loss)
+
+        if reduction == "sum":
+            loss = {"torch.nn.L1Loss": sum(loss_list)}
+        elif reduction == "none":
+            loss = {"torch.nn.L1Loss": loss_list}
+        else:
+            raise ValueError("Reduction not recognised.")
+
+        return loss, inputs_t, image_tensor_list_grad
+
+    def loss_gradient(  # pylint: disable=W0613
+        self, x: np.ndarray, y: Union[List[Dict[str, np.ndarray]], List[Dict[str, "torch.Tensor"]]], **kwargs
+    ) -> np.ndarray:
+        """
+        Compute the gradient of the loss function w.r.t. `x`.
+
+        :param x: Samples of shape (nb_samples, height, width, nb_channels).
+        :param y: Target values of format `List[Dict[Tensor]]`, one for each input image. The
+                  fields of the Dict are as follows:
+
+                  - boxes (FloatTensor[N, 4]): the predicted boxes in [x1, y1, x2, y2] format, with values \
+                    between 0 and H and 0 and W
+                  - labels (Int64Tensor[N]): the predicted labels for each image
+                  - scores (Tensor[N]): the scores or each prediction.
+        :return: Loss gradients of the same shape as `x`.
+        """
+        grad_list = list()
+
+        for i in range(x.shape[0]):
+
+            x_i = x[[i]]
+            y_i = [y[i]]
+
+            output, _, image_tensor_list_grad = self._get_losses(x=x_i, y=y_i)
+
+            # Compute the gradient and return
+            loss = None
+            for loss_name in self.attack_losses:
+                if loss is None:
+                    loss = output[loss_name]
+                else:
+                    loss = loss + output[loss_name]
+
+            # Clean gradients
+            self._model.zero_grad()
+
+            # Compute gradients
+            loss.backward(retain_graph=True)  # type: ignore
+
+            for img in image_tensor_list_grad:
+                gradients = img.grad.cpu().numpy().copy()
+                grad_list.append(gradients)
+
+        grads = np.array(grad_list)
+
+        if grads.shape[0] == 1:
+            grads_ = np.empty(len(grads), dtype=object)
+            grads_[:] = list(grads)
+            grads = grads_
+
+        if not self.all_framework_preprocessing:
+            grads = self._apply_preprocessing_gradient(x, grads)
+
+        if x.dtype != object:
+            grads = np.array([i for i in grads], dtype=x.dtype)  # pylint: disable=R1721
+            assert grads.shape == x.shape and grads.dtype == x.dtype
+
+        return grads
+
+    def _preprocess(self, img):
+        """
+        preprocess image before forward pass, this is the same
+        preprocessing used during training, please refer to collate function
+        in train.py for reference
+        @image: input image
```

Comment:
Can you reformat lines 322-325 consistently with ART style?

Reply:
Ok.